crates/ra_hir/src/adt.rs

```diff
@@ -0,0 +1,213 @@
+    module::Module,
+    ty::{Ty},
+};
+
+pub struct Struct {
+    def_id: DefId,
+}
+
+impl Struct {
+    pub(crate) fn new(def_id: DefId) -> Self {
+        Struct { def_id }
+    }
+
+    pub fn def_id(&self) -> DefId {
+        self.def_id
+    }
+
+    pub fn variant_data(&self, db: &impl HirDatabase) -> Cancelable<Arc<VariantData>> {
+        Ok(db.struct_data(self.def_id)?.variant_data.clone())
+    }
+
+    pub fn struct_data(&self, db: &impl HirDatabase) -> Cancelable<Arc<StructData>> {
+        Ok(db.struct_data(self.def_id)?)
+    }
+
+    pub fn name(&self, db: &impl HirDatabase) -> Cancelable<SmolStr> {
+        Ok(db.struct_data(self.def_id)?.name.clone())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct StructData {
+    name: SmolStr,
+    variant_data: Arc<VariantData>,
+}
+
+impl StructData {
+    pub(crate) fn new(
+        db: &impl HirDatabase,
+        module: &Module,
+        struct_def: ast::StructDef,
+    ) -> Cancelable<StructData> {
+        let name = struct_def
+            .name()
+            .map(|n| n.text())
+            .unwrap_or(SmolStr::new("[error]"));
+        let variant_data = VariantData::new(db, module, struct_def.flavor())?;
+        let variant_data = Arc::new(variant_data);
+        Ok(StructData { name, variant_data })
+    }
+
+    pub fn name(&self) -> &SmolStr {
+        &self.name
+    }
+
+    pub fn variant_data(&self) -> &Arc<VariantData> {
+        &self.variant_data
+    }
+}
+
+pub struct Enum {
+    def_id: DefId,
+}
+
+impl Enum {
+    pub(crate) fn new(def_id: DefId) -> Self {
+        Enum { def_id }
+    }
+
+    pub fn def_id(&self) -> DefId {
+        self.def_id
+    }
+
+    pub fn name(&self, db: &impl HirDatabase) -> Cancelable<SmolStr> {
+        Ok(db.enum_data(self.def_id)?.name.clone())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EnumData {
+    name: SmolStr,
+    variants: Vec<(SmolStr, Arc<VariantData>)>,
+}
+
+impl EnumData {
+    pub(crate) fn new(
+        db: &impl HirDatabase,
+        module: &Module,
+        enum_def: ast::EnumDef,
+    ) -> Cancelable<Self> {
+        let name = enum_def
+            .name()
+            .map(|n| n.text())
+            .unwrap_or(SmolStr::new("[error]"));
+        let variants = if let Some(evl) = enum_def.variant_list() {
+            evl.variants()
+                .map(|v| {
+                    Ok((
+                        v.name()
+                            .map(|n| n.text())
+                            .unwrap_or_else(|| SmolStr::new("[error]")),
+                        Arc::new(VariantData::new(db, module, v.flavor())?),
+                    ))
+                })
+                .collect::<Cancelable<_>>()?
+        } else {
+            Vec::new()
+        };
+        Ok(EnumData { name, variants })
+    }
+}
+
+/// A single field of an enum variant or struct
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct StructField {
+    name: SmolStr,
+    ty: Ty,
```

Comment:
This will eagerly compute the `Ty` of the field. I think it's not a big deal, but perhaps it's better long-term if by default everything is lazy? That is, getting the set of names of fields is one query, getting the type of a specific field is another query?

Reply:
Yeah, I was considering introducing something like `hir::Ty` (we'll need it when introducing `hir::Expr` at the latest). I was then wondering whether that should already be resolved; rustc's HIR already contains the resolved defs, but the resolution is the only potentially expensive thing about constructing a `ty::Ty` at the moment. So I thought I'd leave it for now.